[Boost] Add CSS/JS minification

dilirity opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Add all [Page Optimize](https://github.com/automattic/page-optimize) features under a new feature in Boost - Minify.
* Feature can be toggled on or off from the Boost UI.
**Note:** The feature is called Minify, but only supports what Page Optimize supports - concatenation for CSS and JS. It also supports minification of CSS, but that gets enabled with a PHP constant.
![image](https://user-images.githubusercontent.com/11799079/217846550-2d3252bb-fc1a-4392-ad8b-f5683c896173.png)
* Page Optimize had a settings page under `Settings -> Performance`. That has been removed. There's a discussion around the UI (pc9hqz-1zN-p2) about how we're going to handle the settings transition.
* For the time being, existing settings from Page Optimize get migrated with new names. Migration is not ideal, should be changed later.
* Added GET param (`jb-disable-minify`) to disable Minify to make testing easier.
* The Minify feature and the Page Optimize plugin can be active at the same time, though it's not advised, as they share the same code and it will get executed twice.
* Since Page Optimize relied on PHP constants, code has been written with the idea, that users might have updated settings and this feature [aims to respect them](https://github.com/Automattic/jetpack/blob/5ffc0432043ed166f7f3e2f6d0daa166896524a5/projects/plugins/boost/app/features/optimizations/minify/Config.php).

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

pc9hqz-1yh-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1. Install this version of Boost;
2. Allow the feature on the website, by adding this to `wp-config.php` - `define( 'JETPACK_BOOST_MINIFY', true );`;
3. Enable the Minify feature;
4. Open a page and check source for `_static`. This is the URL the feature uses when serving its cache;
![image](https://user-images.githubusercontent.com/11799079/217850333-2bede350-0fc5-44e6-b637-78b301157df8.png)

5. Enable `WP_DEBUG` to see the handles of the scripts/styles included in the cached url.
![image](https://user-images.githubusercontent.com/11799079/217850116-a501c28b-34c4-4273-8a36-7ddca3b5f311.png)
You might also notice HTML comments - that's the debug output from the feature (shows up only if `WP_DEBUG` is set to `true`).
![image](https://user-images.githubusercontent.com/11799079/217851726-e972cb8a-4ba2-4bb6-bfc6-510ecac752a4.png)

dilirity committed:
Add page optimize plugin files;
Exclude page optimize plugin files from phpcs

dilirity committed:
Allow page optimize to be enabled/disabled via Boost settings;
Update plugin hooks to initialize on boost plugin file;
Prevent minification to happen in WP admin

dilirity committed:
changelog

dilirity committed:
Prefix page optimize code to allow running with original page optimize

dilirity committed:
Update options to avoid collision with official page optimize plugin

dilirity committed:
Add migration for old page optimize options

This is necessary, to make sure the page optimize in Boost is backwards
compatible with existing page optimize installations.

dilirity committed:
Fix fatal error when accessing a cached URL

dilirity committed:
Update concatenated stylesheets to support style_loader_tag filter

dilirity committed:
Allow disabling minify via GET param if enabled

dilirity committed:
Refactor page optimize code as a Boost feature;
Ignore phpcs for minify feature

thingalon commented on projects/plugins/boost/changelog/add-boost-minification:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
```
Getting a bit sick of this setting off the changelogger validity checker. So I'm kicking off a conversation about this. pdWQjU-eR-p2

thingalon replied:
After a discussion with the Crew team, I think we should follow suit and update versions as recommended by the changelogger tool.

pdWQjU-eR-p2

dilirity replied:
I've updated the version in f2dddc57ea2819c06e03d55b9699a33fe18b2981.

dilirity committed:
Handle PHP constants from page optimize;
Add missing functionality from page optimize

dilirity committed:
Remove global variable usage

dilirity committed:
Fix fatal error

dilirity committed:
Fix concatenation initialized too early

dilirity committed:
Remove old page optimize plugin

dilirity committed:
Merge branch 'trunk' into add/boost/minification

dilirity committed:
Update Minify description in UI

dilirity committed:
Fix fatal error during CSS concatenation

Commit:
Merge branch 'trunk' into add/boost/minification

thingalon commented on projects/plugins/boost/compatibility/page-optimize.php:
```diff
@@ -0,0 +1,51 @@
+		return $pre_option;
+	}
+
+	// Prevent recursion.
+	remove_filter( 'pre_option', 'jetpack_boost_page_optimize_migrate_options', 10, 2 );
+
+	// Populate new options from old options.
+	foreach ( $legacy_options as $legacy_option_name ) {
+		$old_value = get_option( $legacy_option_name );
+		if ( false !== $old_value ) {
+			update_option( 'jetpack_boost_' . $legacy_option_name, $old_value );
+			delete_option( $legacy_option_name );
+
+			// Update current option value, to use correct value (old one).
+			if ( $legacy_option_name === $current_legacy_option_name ) {
+				$pre_option = $old_value;
+			}
+		}
+	}
```
If I'm reading this correctly, the first time anything tries to read any of the options related to pageoptimize, it will rename all pageoptimize settings to `jetpack_boost_` versions of themselves. Then, on subsequent pageloads, it will try to do the same thing again.

A few questions:
1. Do we need to rename these options? Would it make more sense to leave them unmodified?
2. If we are going to rename these options, why not do it when the module is first activated, once?

I feel like this will cause a lot of unnecessary options reads on future page-loads.

Additionally, this php file is loaded only if PageOptimize is also enabled. What happens if a user turns off PageOptimize, then turns on Boost?

dilirity replied:
As discussed in our chat - we won't be doing any option migration for the time being.

I've removed the migration code (5e203dc6970a9232135e4fa56863c73ebf5db868) and updated it to use the original page optimize option names in d3170452ade5bb43a6ca97855a1215bbd7dd7c19. To test things out - I activated Page Optimize, disabled JS concatenation and it stopped concatenating. Enabled it after that - it continued to work.

Edit: just realized I could've named 5e203dc6970a9232135e4fa56863c73ebf5db868 better...

thingalon commented on projects/plugins/boost/app/features/optimizations/minify/Minify.php:
```diff
@@ -0,0 +1,66 @@
+		}
+
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		$disable = isset( $_GET['jb-disable-minify'] ) || jetpack_boost_page_optimize_bail();
```
There is already a GET parameter; `jb-disable-modules` which lets you pass in a list of module slugs to disable (or `all` for all). Do we need a new GET param?

dilirity replied:
I wasn't aware of that GET param - thanks.

Updated with [f477e23](https://github.com/Automattic/jetpack/pull/28686/commits/f477e237a3c15971782f0ef61930ca091bf3ca13).

thingalon commented on projects/plugins/boost/app/features/optimizations/minify/cssmin/Colors.php:
```diff
@@ -0,0 +1,155 @@
+			'lightsteelblue' => '#b0c4de',
+			'lightyellow' => '#ffffe0',
+			'limegreen' => '#32cd32',
+			'mediumaquamarine' => '#66cdaa',
+			'mediumblue' => '#0000cd',
+			'mediumorchid' => '#ba55d3',
+			'mediumpurple' => '#9370db',
+			'mediumseagreen' => '#3cb371',
+			'mediumslateblue' => '#7b68ee',
+			'mediumspringgreen' => '#00fa9a',
+			'mediumturquoise' => '#48d1cc',
+			'mediumvioletred' => '#c71585',
+			'midnightblue' => '#191970',
```
🤣 things I never considered in minification, but it makes sense.

dilirity replied:
Huh, that is so neat I love it :D

The [reverse one](https://github.com/Automattic/jetpack/pull/28686/files/b7facb58c94a5da9124b223fe91ffa1627dad567#diff-89574fcbffcf9d5bc264e9c2b0f2d5e081d2533a8006665532e1b4d870c68b1eR11-R17) is also fun.

thingalon commented on projects/plugins/boost/app/features/optimizations/minify/cssmin/cssmin.php:
```diff
@@ -0,0 +1,8 @@
+<?php
+
+// Simple loader file
+// Library downloaded directly from https://github.com/tubalmartin/YUI-CSS-compressor-PHP-port
```
cssmin appears to use the [BSD license](https://github.com/rgrove/cssmin/blob/master/LICENSE), which requires the license be attached. I don't see the BSD license text for it in here?

Can we install cssmin via composer, to save adding it to our repository?

dilirity replied:
Yep yep! Done - 24b679cde2a4af547ccadd7aa1166bd0224aa34d.

thingalon reviewed (changes requested):
Can we please have a doc-block on all functions over about 10 lines long explaining their purpose.

Got some thoughts, questions and comments inline:

dilirity committed:
Remove page optimize compatibility

Options won't be migrated. An upcoming commit will revert to the
option names page optimize uses.

dilirity committed:
Update page optimize settings to use original names

dilirity committed:
Remove ability to disable minify feature via specific GET param

There's already a param for this - jb-disable-modules.

dilirity committed:
Update cssmin to use composer package instead of separate files

thingalon commented:
I'm seeing some weird behaviour. e.g.:

### Messed up ids

Before: 
![image](https://user-images.githubusercontent.com/1369626/218645505-de3cd2fc-b9a8-4789-9b06-6b0816549c8c.png)

After:
![image](https://user-images.githubusercontent.com/1369626/218646379-bc508ea4-b99b-4e99-919e-94d9b124168c.png)

That id doesn't look right. Is that expected behaviour?

### Unminified content
![image](https://user-images.githubusercontent.com/1369626/218646071-9bbbd2a3-89fc-42af-8b51-d3d70dc6b75e.png)

Should that be minified? I am using the default settings. Is this version concat-only?

dilirity commented:
@thingalon by default the plugin provides CSS and JS concatenation. CSS minification is off by default and is something that has to be turned on via a PHP constant.

As for the id - that's something that carried over from page optimize.

It kind of slipped through the cracks. Not sure what id to add though, since the original is added to a `data-handle` (though not the complete one - missing `-css`).

I'll see what I can cook up.

EDIT:
Just realized that data-handle contains the handle of the asset, not the id itself.

dilirity committed:
Fix versioning

dilirity committed:
Hide minify feature behind a PHP constant

dilirity commented:
The Minify module will now be disabled by default. To enable, add this to `wp-config.php`:
```php
define( 'JETPACK_BOOST_MINIFY', true );
```

Please note, that this constant does not enable the minification. It enables including the code for it. To enable the feature, you'll still need to enable it via the UI.

thingalon reviewed (approved):
👍 With that gating, I am happy for this to get merged and the issues identified (which came bundled with the PageOptimize source) can be tackled in separate PRs.

@dilirity can you please make tickets for the remaining issues after this is merged, and put the in the current focus board?
